### DISCUSSION TITLE
Drag and drop multiple models in the editor for composition

### DISCUSSION THREAD

**Discussion by hsd-dev** [eclipse-glsp/glsp#726] [Q&A]
I have a model similar to the [coffee machine model](https://eclipsesource.com/blogs/2020/07/24/a-web-based-modeling-tool-based-on-eclipse-theia/) (described in `The Underlying Model`) 
```
Components can contain other components
```
The difference being that each component is defined as an independent component in a separate file which can be composed together in a higher level component in a new file. I have been able to achieve this fairly well in a Xtext based editor (still a WIP though). Would this also be possible in GLSP? 

My idea would be to make the lower level components visible in the tool palette and drag and drop them.

**Comment by planger** [comment]
Yes, that should be possible. To the client it doesn't matter how your diagram is composed. It just receives a GModel in the end. Also the command palette items are requested from and provided the server (see e.g. [Workflow example](https://github.com/eclipse-glsp/glsp-server/blob/master/examples/org.eclipse.glsp.example.workflow/src/org/eclipse/glsp/example/workflow/provider/WorkflowCommandPaletteActionProvider.java)).

Thus, in the end all your custom logic for composition needs to be implemented on the server. That is, it will have to read the available lower-level components and offer them as palette items.
Once such a lower-level component is added on the client, the server will receive the respective operation (depends on what action you use, but you could implement that with a `CreateNodeOperation` with custom arguments I guess). On such an operation, the server would modify the model state in a way that eventually the GModel that is being created also contains the diagram for the lower-level components.

**Reply by hsd-dev** [reply]
@planger thanks for the quick response and the solution makes sense! I will keep the discussion open to track the progress and ask related questions here.
